spidermonkey: update to 102 and fix build under macOS

catap opened the pull request:
###### Description of changes

This PR adds `spidermonkey_102` which can be used on every unix which includes macOS. It also fix build of old version under macOS.

Thus, it suppress / consumes / closes #172376 and fixes #91415

And it adds tests for all packages :)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SuperSandro2000 commented:
@ofborg eval

SuperSandro2000 commented on pkgs/development/interpreters/spidermonkey/adds-Wno-deprecated-declarations-icu.patch:
```diff
@@ -0,0 +1,30 @@
+         '-Wno-tautological-unsigned-enum-zero-compare',
+         '-Wno-unreachable-code-loop-increment',
+         '-Wno-unreachable-code-return',
++        '-Wno-deprecated-declarations',
```
Isn't this warning only?

catap replied:
it builds with any warning is an error => it fails.

catap replied:
Let me double check it. Maybe it failed with different reason

SuperSandro2000 commented on pkgs/development/interpreters/spidermonkey/102.nix:
```diff
@@ -0,0 +1,129 @@
+{ callPackage
```
Please create a common or generic file.

SuperSandro2000 commented on pkgs/development/interpreters/spidermonkey/102.nix:
```diff
@@ -0,0 +1,129 @@
+
+  patches = [
+    ./adds-Wno-deprecated-declarations-icu.patch
+  ] ++ lib.optionals stdenv.isDarwin [
```
If the file does not break linux build, we should always apply it.

catap commented:
@SuperSandro2000 may I ask you to re-review this one also? Thanks!

jtojnar commented:
It would be nice to split the changes into independent commits, for example:

- spidermonkey_*: Factor out into a common expression
- spidermonkey_*: Fix build under macOS
- spidermonkey_102: init at 102.0

That will make it easier to review.

catap commented:
@jtojnar I've done it in different way:
1. added 102 with macOS support
2. added macOS support for 91 and 78;
3. refactored it.

jtojnar commented:
Having (2) separate from the other changes is the most important. But having the unification first would be the best since it minimizes the number of files needed to compare.

catap commented:
@jtojnar ok, let me try do it. It shouldn't be something complicated.

catap committed:
spidermonkey_102: init at 102.0

catap committed:
spidermonkey_*: add myself as a maintainer

catap committed:
spidermonkey_91: factor out into a common expression

catap committed:
spidermonkey_78: factor out into a common expression

catap committed:
spidermonkey_*: add tests

catap committed:
spidermonkey_*: remove useless dependency

catap commented:
@jtojnar here it is.

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,179 @@
+    m4
+    perl
+    pkg-config
+    python39 # 78 requires python up to 3.9
```
Avoid introducing extra Python dependency into the build closure on systems that use new Spidermonkey just because the old one does not build with it:

```suggestion
    # 78 requires python up to 3.9
    (if lib.versionOlder version "91" then python3 else python39)
```

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,179 @@
+    zip
+  ] ++ lib.optionals (lib.versionOlder version "91") [
+    autoconf213
+    yasm
```
This had a comment `# to buid icu? seems weird` in 78.

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,127 @@
+  meta = with lib; {
+    description = "Mozilla's JavaScript engine written in C/C++";
+    homepage = "https://spidermonkey.dev/";
+    license = licenses.mpl20;
```
This had a comment `# TODO: MPL/GPL/LGPL tri-license.` in 78.

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -55,6 +61,24 @@ stdenv.mkDerivation rec {
     zlib
   ];
 
+  patches = lib.optional (lib.versionOlder version "91") [
```
The 78 expression had an ordering of attributes `patches` just after `src`, which is slightly more common, see https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-ordering.md. It might be nice to preserve it.

jtojnar replied:
This should be `lib.optionals` otherwise you get list of lists.

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,127 @@
+    "--with-system-icu"
+    "--with-system-nspr"
+    "--with-system-zlib"
+    "--enable-optimize"
```
78 included a bit more context:

```
    # Fedora and Arch disable optimize, but it doesn't seme to be necessary
    # It turns on -O3 which some gcc version had a problem with:
    # https://src.fedoraproject.org/rpms/mozjs38/c/761399aba092bcb1299bb4fccfd60f370ab4216e
```

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,127 @@
+    "--enable-readline"
+    "--enable-release"
+    "--enable-shared-js"
+    "--disable-debug"
+    "--disable-jemalloc"
+    "--disable-strip"
+    "--disable-tests"
```
These three were not enabled in 78:

```
    "--disable-debug"
    "--disable-strip"
    "--disable-tests"
```

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -51,16 +54,24 @@ stdenv.mkDerivation (finalAttrs: rec {
     yasm
+  ] ++ lib.optionals stdenv.isDarwin [
+    xcbuild
+    libobjc
+    libiconv
```
Do these really go here rather than `buildInputs`?

catap replied:
@jtojnar yes, it is linked against `libiconv.dylib` on macOS:
```sh
√ result % otool -L lib/libmozjs-102.dylib     
lib/libmozjs-102.dylib:
	/nix/store/9ld6s0lcdjw9cp26zf4qa0j24m80rwd5-spidermonkey-102.0/lib/libmozjs-102.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
	/nix/store/dpfzpp8c7z0fdxxvin21pk5gs135wn39-Libsystem-1238.60.2/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/nix/store/k56igcml51xccxxakgvj8lai6f7fz05q-zlib-1.2.12/lib/libz.dylib (compatibility version 1.0.0, current version 1.2.12)
	/nix/store/q4c16zw3kimqdvxcc06njrv49bq68hpf-libiconv-50/lib/libiconv.dylib (compatibility version 7.0.0, current version 7.0.0)
	/nix/store/nbwbggjf37pxmishd2sqqbswv6f0l3cb-libcxx-11.1.0/lib/libc++.1.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1454.90.0)
√ result % otool -L bin/js102     
bin/js102:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
	/nix/store/dpfzpp8c7z0fdxxvin21pk5gs135wn39-Libsystem-1238.60.2/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/nix/store/k56igcml51xccxxakgvj8lai6f7fz05q-zlib-1.2.12/lib/libz.dylib (compatibility version 1.0.0, current version 1.2.12)
	/nix/store/qgzlayc4acky2pyp56dh6p1bm00d315g-readline-6.3p08/lib/libreadline.6.dylib (compatibility version 6.0.0, current version 6.3.0)
	/nix/store/q4c16zw3kimqdvxcc06njrv49bq68hpf-libiconv-50/lib/libiconv.dylib (compatibility version 7.0.0, current version 7.0.0)
	/nix/store/nbwbggjf37pxmishd2sqqbswv6f0l3cb-libcxx-11.1.0/lib/libc++.1.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1454.90.0)
√ result % 
```

SuperSandro2000 replied:
If it is linked against a library it goes into buildInputs.

SuperSandro2000 replied:
I think this is the last open thread.

```suggestion
    libobjc
```
libiconv is already in buildInputs. Can we mopve libobjc, too?

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -160,6 +174,6 @@ stdenv.mkDerivation (finalAttrs: rec {
     license = licenses.mpl20;
     maintainers = with maintainers; [ abbradar lostnet catap ];
-    platforms = platforms.linux;
+    platforms = platforms.unix;
```
Does this work on Darwin for all versions or just 102? The conditional expressions above seem to hint at the latter.

catap replied:
@jtojnar yes, it does. See included tests which was successfully executed by ofborg

jtojnar reviewed (commented):
Thanks, this is now much easier to review.

catap commented:
@jtojnar seems that I've addressed all your comments. Or at least I hope so.

To make review easy I haven't rebased anything, and keep all changes as two dedicated commits.

lostnet commented:
It seems to me like the removed adds-Wno-deprecated-declarations-icu.patch may have been a fix of system icu for 102esr on darwin?

If the patch did work my preference from a couchdb perspective is to use the system icu where possible since couchdb independently depends on the system icu, but that's not a very high priority.

catap commented:
@lostnet things a bit more complicated. It is failed since https://github.com/mozilla/gecko-dev/commit/9aa3587bbf0416dd2eb5b614f7b301c71c64286b 

The patch wasn't needed at all, it was a wrong lead which I forgot to remove.

I may hack build system a bit more to allow using both system's icu and nspr, but it seems quite ugly.

mweinelt commented:
Passing `--with-system-icu and `--with-system-nspr` should be possible. We do use that on `buildMozillaMach`, also for 91 and 102 ESR releases. But we also only build for the two Linux targets.

catap commented:
@mweinelt are you sure? I see that firefox 102 has `badPlatforms = lib.platforms.darwin`: https://github.com/NixOS/nixpkgs/blob/be948cdf737fec2f1d69ff721da2c0bd2232bb46/pkgs/applications/networking/browsers/firefox/packages.nix#L43

mweinelt commented:
I'm not sure what about Darwin is broken. It hasn't been maintained for more than 3y, and even then it was on and off.

What error are you seeing with both enabled?

catap commented:
@mweinelt it simple fails on configure stage because icu / nsr can't be used from system if it bootstraped from scratch.

catap commented:
@mweinelt to summarize: `--with-system-icu` and `--with-system-nspr` aren't used only on macOS and only for 102+ version.

jtojnar commented:
Its nice to preserve the commits for people doing code archaeology with `git blame`.

catap commented:
@jtojnar when let keep it on this way

catap commented:
I've:
1. revased to last master and consume fixup commits;
2. added a trivial patch to allow using system's icu and nspr on macOS.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/548

catap committed:
spidermonkey_*: fix build under macOS

catap committed:
spidermonkey_*: follow guideline for expression ordering

lostnet commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,184 @@
+, which
+, zip
+, autoconf213
+, yasm
```
I noticed with conda's port that they didn't seem to need yasm so it might be safe to remove at least in 91+,

catap replied:
@lostnet it is used only at 78 ;)

lostnet reviewed (approved):
I've looked over the refactor, mac support+patches, and passthru test and it all looks good to me.

catap commented:
@lostnet any idea what should I do to have it merged?

ajs124 reviewed (approved):
LGTM. Since I haven't read through all the discussion, were all relevant reviews (e.g. by @jtojnar) addressed?

siraben commented:
Result of `nixpkgs-review pr 179822` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>spidermonkey</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>jsawk</li>
    <li>spidermonkey_78</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>couchdb3</li>
    <li>spidermonkey_102</li>
    <li>spidermonkey_91</li>
  </ul>
</details>

catap commented:
@siraben did it builds before?

catap commented:
Rebased to the last master to fix conflict with https://github.com/NixOS/nixpkgs/pull/182849

@mweinelt maybe we should merge it?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/567

SuperSandro2000 commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,184 @@
+    m4
+    perl
+    pkg-config
+    # 78 requires python up to 3.9
+    (if lib.versionOlder version "91" then python39 else python3)
```
Please convert that into the style used already in line 79.

SuperSandro2000 commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,184 @@
+  ];
+
+  buildInputs = [
+    (if lib.versionOlder version "91" then icu67 else icu)
```
Please use the style already used in patches.

SuperSandro2000 commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,184 @@
+    # We can't build in js/src/, so create a build dir
+    mkdir obj
+    cd obj/
+    configureScript=../js/src/configure
```
```suggestion
    export configureScript=../js/src/configure
```
Shouldn't this be set in nix?

SuperSandro2000 commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,184 @@
+    zlib
+  ];
+
+  preConfigure = ''
+    ${lib.optionalString (lib.versionOlder version "91") ''
+    export CXXFLAGS="-fpermissive"
+    ''}
+    export LIBXUL_DIST=$out
+    export PYTHON="${buildPackages.python3.interpreter}"
+
+    ${lib.optionalString (lib.versionAtLeast version "91") ''
+    export M4=m4
+    export AWK=awk
+    export AS=$CC
+    export AC_MACRODIR=$PWD/build/autoconf/
+
+    pushd js/src
+    sh ../../build/autoconf/autoconf.sh --localdir=$PWD configure.in > configure
+    chmod +x configure
+    popd
+    ''}
+
```
```suggestion
  preConfigure = lib.optionalString (lib.versionOlder version "91") ''
    export CXXFLAGS="-fpermissive"
  '' + ''
    export LIBXUL_DIST=$out
    export PYTHON="${buildPackages.python3.interpreter}"
  '' + lib.optionalString (lib.versionAtLeast version "91") ''
    export M4=m4
    export AWK=awk
    export AS=$CC
    export AC_MACRODIR=$PWD/build/autoconf/

    pushd js/src
    bash ../../build/autoconf/autoconf.sh --localdir=$PWD configure.in > configure
    chmod +x configure
    popd
  '' + ''
```

SuperSandro2000 reviewed (commented):
otherwise LGTM

catap committed:
spidermonkey_*: follow review remarks

catap commented:
@SuperSandro2000 may I ask you to re-review this one?

wegank commented:
@catap [This comment](https://github.com/NixOS/nixpkgs/pull/179822#discussion_r935691129) has yet to be solved.

wegank commented:
Just saw that [GNOME 43 will depend on SpiderMonkey 102](https://discourse.gnome.org/t/gnome-43-to-depend-on-spidermonkey-102/10658)!

catap commented:
@wegank thanks for the ping. I'll take a look on this on couple of hours.

jtojnar commented on pkgs/development/interpreters/spidermonkey/common.nix:
```diff
@@ -0,0 +1,188 @@
+  patches = lib.optionals (lib.versionAtLeast version "102") [
+    # use pkg-config at all systems
+    ./always-check-for-pkg-config.patch
+    ./allow-system-s-nspr-and-icu-on-bootstrapped-sysroot.patch
```
```suggestion
    ./allow-system-s-nspr-and-icu-on-bootstrapped-sysroot.patch
    
    # Patches required by GJS
    # https://discourse.gnome.org/t/gnome-43-to-depend-on-spidermonkey-102/10658

    # Install ProfilingCategoryList.h
    (fetchpatch {
      url = "https://hg.mozilla.org/releases/mozilla-esr102/raw-rev/33147b91e42b79f4c6dd3ec11cce96746018407a";
      sha256 = "sha256-xJFJZMYJ6P11HQDZbr48GFgybpAeVcu3oLIFEyyMjBI=";
    })

    # Fix embeder build
    (fetchpatch {
      url = "https://hg.mozilla.org/releases/mozilla-esr102/raw-rev/1fa20fb474f5d149cc32d98df169dee5e6e6861b";
      sha256 = "sha256-eCisKjNxy9SLr9KoEE2UB26BflUknnR7PIvnpezsZeA=";
    })
```

jtojnar commented on pkgs/development/interpreters/spidermonkey/102.nix:
```diff
@@ -0,0 +1,4 @@
+import ./common.nix {
+  version = "102.0";
```
102.1.0 may have been released according to https://discourse.gnome.org/t/gnome-43-to-depend-on-spidermonkey-102/10658.

jtojnar commented:
Since this is blocking #182618, I cleaned up it a bit, addressed the remaining review comments and reverted some pointless changes (e.g. exporting `configureScript`): https://github.com/NixOS/nixpkgs/pull/185928
